File: apps/zap-downloader-node/src/commands/daemon.ts
import * as path from 'path';
import * as fs from 'fs';
import chalk from 'chalk';
import { Arguments } from 'yargs';
import pm2 from 'pm2';
import * as toml from '@iarna/toml';
import { getWorkspace } from '../workspace';

type PM2ProcessInfo = any;

interface TomlConfig {
  ENV?: {
    ZAP_DOWNLOADER_WORKSPACE?: string;
    ZAP_DOWNLOADER_DOWNLOADS?: string;
    ZAP_DOWNLOADER_INSTALL?: string;
    ZAP_DOWNLOADER_PACKAGES?: string;
    ZAP_DOWNLOADER_ZAP_HOME?: string;
  };
  SERVER?: {
    PORT?: number;
    HOST?: string;
  };
  PATHS?: {
    JAR_PATH?: string;
    DIR?: string;
    INSTALL_DIR?: string;
  };
  JAVA_OPTIONS?: {
    flags?: string[];
  };
  CONFIG?: {
    flags?: string[];
  };
}

export const command = 'daemon';
export const describe = 'Manage ZAP daemon';

export const builder = (yargs: any) => {
  return yargs
    .command(startDaemonCommand)
    .command(stopDaemonCommand)
    .demandCommand(1, 'Specify a subcommand: start-daemon, stop-daemon');
};

export const handler = () => {};

function parseTomlConfig(tomlPath: string): TomlConfig {
  const content = fs.readFileSync(tomlPath, 'utf-8');
  return toml.parse(content) as TomlConfig;
}

function resolveTomlPaths(config: TomlConfig, defaultWorkspace: string): TomlConfig {
  const workspace = config.ENV?.ZAP_DOWNLOADER_WORKSPACE || defaultWorkspace;
  
  return {
    ...config,
    ENV: {
      ...config.ENV,
      ZAP_DOWNLOADER_WORKSPACE: workspace,
    },
    PATHS: {
      ...config.PATHS,
      JAR_PATH: config.PATHS?.JAR_PATH || '',
      DIR: config.PATHS?.DIR || '.zap',
      INSTALL_DIR: config.PATHS?.INSTALL_DIR || path.join(workspace, 'install').replace(/\\/g, '/'),
    },
  };
}

const startDaemonCommand = {
  command: 'start-daemon',
  describe: 'Start ZAP as a daemon using pm2',
  builder: (yargs: any) => {
    return yargs
      .option('toml', {
        alias: 't',
        description: 'Path to zap.toml configuration file',
        type: 'string',
      })
      .option('dir', {
        alias: 'd',
        description: 'ZAP installation directory (where zap.jar is)',
        type: 'string',
      })
      .option('workspace', {
        alias: 'w',
        description: 'ZAP working directory',
        type: 'string',
      })
      .option('host', {
        description: 'ZAP host to bind to',
        type: 'string',
        default: '0.0.0.0',
      })
      .option('port', {
        alias: 'P',
        description: 'ZAP proxy port',
        type: 'number',
        default: 8080,
      })
      .option('api-key', {
        alias: 'k',
        description: 'ZAP API key',
        type: 'string',
        default: '',
      })
      .option('name', {
        alias: 'N',
        description: 'PM2 process name',
        type: 'string',
        default: 'zap-daemon',
      });
  },
  handler: async (argv: Arguments & {
    toml?: string;
    dir?: string;
    workspace?: string;
    host?: string;
    port?: number;
    apiKey?: string;
    name?: string;
  }) => {
    let config: TomlConfig = {};
    let useToml = false;
    let workspace = argv.workspace || getWorkspace();

    if (argv.toml) {
      if (!fs.existsSync(argv.toml)) {
        console.error(chalk.red(`TOML file not found: ${argv.toml}`));
        process.exit(1);
      }
      console.log(chalk.blue(`Using TOML config: ${argv.toml}`));
      config = parseTomlConfig(argv.toml);
      config = resolveTomlPaths(config, workspace);
      useToml = true;
    }

    const host = useToml ? (config.SERVER?.HOST || '0.0.0.0') : (argv.host || '0.0.0.0');
    const port = useToml ? (config.SERVER?.PORT || 8080) : (argv.port || 8080);
    const apiKey = useToml ? '' : (argv.apiKey || '');
    const processName = argv.name || 'zap-daemon';
    
    const zapHomeDir = useToml 
      ? (config.ENV?.ZAP_DOWNLOADER_ZAP_HOME || '.zap')
      : '.zap';

    const zapInstallDir = useToml
      ? (config.PATHS?.INSTALL_DIR || path.join(workspace, 'install'))
      : (argv.dir || path.join(argv.workspace || getWorkspace(), 'zap'));
    
    const workingDir = useToml
      ? (config.ENV?.ZAP_DOWNLOADER_WORKSPACE || workspace)
      : (argv.workspace || getWorkspace());

    let jarPath: string | null = null;

    if (useToml && config.PATHS?.JAR_PATH) {
      if (fs.existsSync(config.PATHS.JAR_PATH)) {
        jarPath = config.PATHS.JAR_PATH;
      }
    }

    if (!jarPath && fs.existsSync(zapInstallDir)) {
      const files = fs.readdirSync(zapInstallDir);
      for (const f of files) {
        if (f.endsWith('.jar') && f.startsWith('zap')) {
          jarPath = path.join(zapInstallDir, f);
          break;
        }
      }
    }

    if (!jarPath) {
      console.error(chalk.red(`JAR file not found in: ${zapInstallDir}`));
      process.exit(1);
    }

    if (!fs.existsSync(workingDir)) {
      fs.mkdirSync(workingDir, { recursive: true });
    }

    const tmpDir = path.join(workingDir, 'tmp');
    if (!fs.existsSync(tmpDir)) {
      fs.mkdirSync(tmpDir, { recursive: true });
    }

    const zapDir = path.join(workingDir, zapHomeDir);
    if (!fs.existsSync(zapDir)) {
      fs.mkdirSync(zapDir, { recursive: true });
    }

    const pluginDir = path.join(zapDir, 'plugin');
    const installPluginDir = path.join(zapInstallDir, 'plugin');
    if (!fs.existsSync(pluginDir) && fs.existsSync(installPluginDir)) {
      fs.mkdirSync(pluginDir, { recursive: true });
      const pluginFiles = fs.readdirSync(installPluginDir);
      for (const f of pluginFiles) {
        const src = path.join(installPluginDir, f);
        const dest = path.join(pluginDir, f);
        if (!fs.existsSync(dest)) {
          fs.copyFileSync(src, dest);
        }
      }
    }

    const absJarPath = path.resolve(jarPath);
    const absInstallDir = path.dirname(absJarPath);
    const absWorkingDir = path.resolve(workingDir);
    const absTmpDir = path.resolve(tmpDir);
    const absZapDir = path.resolve(zapDir);

    const javaOptions: string[] = [];
    
    if (useToml && config.JAVA_OPTIONS?.flags) {
      javaOptions.push(...config.JAVA_OPTIONS.flags);
    } else {
      javaOptions.push('-Xmx2g');
    }

    javaOptions.push(`-Djava.io.tmpdir="${absTmpDir}"`);

    const configFlags: string[] = [];
    
    if (useToml && config.CONFIG?.flags) {
      configFlags.push(...config.CONFIG.flags);
    } else {
      configFlags.push(
        apiKey ? `api.key=${apiKey}` : 'api.disablekey=true',
        'api.addrs.addr.name=.*',
        'api.addrs.addr.regex=true'
      );
    }

    const javaFlagsStr = javaOptions.join(' ');
    const configFlagsStr = configFlags.map(f => `-config ${f}`).join(' ');

    const cmd = [
      'java',
      javaFlagsStr,
      `-jar "${absJarPath}"`,
      '-daemon',
      `-dir "${absZapDir}"`,
      `-installdir "${absInstallDir}"`,
      `-host ${host}`,
      `-port ${port}`,
      configFlagsStr,
    ].filter(Boolean).join(' ');

    const scriptPath = path.join(workingDir, 'start-zap.sh');
    const scriptContent = `#!/bin/bash
${cmd}
`;
    fs.writeFileSync(scriptPath, scriptContent);

    console.log(chalk.blue(`Starting ZAP daemon with pm2...`));
    console.log(chalk.gray(`JAR: ${jarPath}`));
    console.log(chalk.gray(`Port: ${port}`));
    console.log(chalk.gray(`Working dir: ${workingDir}`));
    if (useToml) {
      console.log(chalk.gray(`Config: TOML (${argv.toml})`));
    }

    try {
      await new Promise<void>((resolve, reject) => {
        pm2.connect((err) => {
          if (err) {
            reject(err);
            return;
          }
          resolve();
        });
      });

      const processes = await new Promise<PM2ProcessInfo[]>((resolve, reject) => {
        pm2.list((err, list) => {
          if (err) reject(err);
          else resolve(list);
        });
      });

      const existing = processes.find(p => p.name === processName);
      if (existing) {
        console.log(chalk.yellow(`Stopping existing pm2 process: ${processName}`));
        await new Promise<void>((resolve, reject) => {
          pm2.stop(processName, (err) => {
            if (err) reject(err);
            else resolve();
          });
        });
        await new Promise<void>((resolve, reject) => {
          pm2.delete(processName, (err) => {
            if (err) reject(err);
            else resolve();
          });
        });
      }

      await new Promise<void>((resolve, reject) => {
        pm2.start({
          name: processName,
          script: 'bash',
          args: scriptPath,
          cwd: workingDir,
        }, (err) => {
          if (err) reject(err);
          else resolve();
        });
      });

      console.log(chalk.green(`ZAP daemon started as "${processName}"`));
      console.log(chalk.blue(`ZAP is starting up on ${host}:${port}`));
    } catch (err: any) {
      console.error(chalk.red(`Failed to start ZAP daemon: ${err.message}`));
      process.exit(1);
    } finally {
      pm2.disconnect();
    }
  },
};

const stopDaemonCommand = {
  command: 'stop-daemon',
  describe: 'Stop ZAP daemon managed by pm2',
  builder: (yargs: any) => {
    return yargs
      .option('name', {
        alias: 'N',
        description: 'PM2 process name',
        type: 'string',
        default: 'zap-daemon',
      });
  },
  handler: async (argv: Arguments & {
    name?: string;
  }) => {
    const processName = argv.name || 'zap-daemon';

    console.log(chalk.blue(`Stopping ZAP daemon: ${processName}...`));

    try {
      await new Promise<void>((resolve, reject) => {
        pm2.connect((err) => {
          if (err) {
            reject(err);
            return;
          }
          resolve();
        });
      });

      const processes = await new Promise<PM2ProcessInfo[]>((resolve, reject) => {
        pm2.list((err, list) => {
          if (err) reject(err);
          else resolve(list);
        });
      });

      const existing = processes.find(p => p.name === processName);
      if (!existing) {
        console.log(chalk.yellow(`No pm2 process found: ${processName}`));
        pm2.disconnect();
        return;
      }

      await new Promise<void>((resolve, reject) => {
        pm2.stop(processName, (err) => {
          if (err) reject(err);
          else resolve();
        });
      });

      await new Promise<void>((resolve, reject) => {
        pm2.delete(processName, (err) => {
          if (err) reject(err);
          else resolve();
        });
      });

      console.log(chalk.green(`ZAP daemon stopped and removed from pm2`));
    } catch (err: any) {
      console.error(chalk.red(`Failed to stop ZAP daemon: ${err.message}`));
      process.exit(1);
    } finally {
      pm2.disconnect();
    }
  },
};

File: apps/zap-downloader/zap_downloader/commands/daemon.py
import typer
import os
import subprocess
import signal
import shutil
import toml
from pathlib import Path
from rich.console import Console
from typing import Optional, Dict, Any, List

console = Console()

PID_FILE = "zap-daemon.pid"
PROCESS_NAME = "zap-daemon"


def parse_toml_config(toml_path: str) -> Dict[str, Any]:
    with open(toml_path, "r") as f:
        return toml.load(f)


def resolve_toml_paths(
    config: Dict[str, Any], default_workspace: str
) -> Dict[str, Any]:
    workspace = config.get("ENV", {}).get("ZAP_DOWNLOADER_WORKSPACE", default_workspace)

    env = config.get("ENV", {})
    env["ZAP_DOWNLOADER_WORKSPACE"] = workspace

    paths = config.get("PATHS", {})
    paths["JAR_PATH"] = paths.get("JAR_PATH", "")
    paths["DIR"] = paths.get("DIR", ".zap")
    paths["INSTALL_DIR"] = paths.get(
        "INSTALL_DIR", os.path.join(workspace, "install").replace("\\", "/")
    )

    return config


def start_daemon(
    toml: Optional[str] = typer.Option(
        None, "--toml", "-t", help="Path to zap.toml configuration file"
    ),
    dir: Optional[str] = typer.Option(
        None, "--dir", "-d", help="ZAP installation directory (where zap.jar is)"
    ),
    workspace: Optional[str] = typer.Option(
        None, "--workspace", "-w", help="ZAP working directory"
    ),
    host: str = typer.Option("0.0.0.0", "--host", help="ZAP host to bind to"),
    port: int = typer.Option(8080, "--port", "-P", help="ZAP proxy port"),
    api_key: str = typer.Option("", "--api-key", "-k", help="ZAP API key"),
    name: str = typer.Option(
        "zap-daemon", "--name", "-N", help="Process name for tracking"
    ),
):
    """Start ZAP as a daemon."""
    from ..workspace import get_workspace

    config: Dict[str, Any] = {}
    use_toml = False

    if toml:
        if not os.path.exists(toml):
            console.print(f"[red]TOML file not found: {toml}[/red]")
            raise typer.Exit(1)
        console.print(f"[blue]Using TOML config: {toml}[/blue]")
        config = parse_toml_config(toml)
        config = resolve_toml_paths(config, workspace or get_workspace())
        use_toml = True

    workspace = workspace or get_workspace()

    if use_toml:
        host = config.get("SERVER", {}).get("HOST", "0.0.0.0")
        port = config.get("SERVER", {}).get("PORT", 8080)
        api_key = ""
        zap_home = config.get("ENV", {}).get("ZAP_DOWNLOADER_ZAP_HOME", ".zap")
        zap_install_dir = config.get("PATHS", {}).get(
            "INSTALL_DIR", os.path.join(workspace, "install")
        )
        workspace = config.get("ENV", {}).get("ZAP_DOWNLOADER_WORKSPACE", workspace)
    else:
        zap_home = ".zap"
        zap_install_dir = dir if dir else os.path.join(workspace, "zap")

    jar_path = None

    if use_toml and config.get("PATHS", {}).get("JAR_PATH"):
        if os.path.exists(config["PATHS"]["JAR_PATH"]):
            jar_path = config["PATHS"]["JAR_PATH"]

    if not jar_path and os.path.exists(zap_install_dir):
        for f in os.listdir(zap_install_dir):
            if f.endswith(".jar") and f.startswith("zap"):
                jar_path = os.path.join(zap_install_dir, f)
                break

    if not jar_path:
        console.print(f"[red]JAR file not found in: {zap_install_dir}[/red]")
        raise typer.Exit(1)

    if not os.path.exists(workspace):
        os.makedirs(workspace, exist_ok=True)

    tmp_dir = os.path.join(workspace, "tmp")
    if not os.path.exists(tmp_dir):
        os.makedirs(tmp_dir, exist_ok=True)

    zap_dir = os.path.join(workspace, zap_home)
    if not os.path.exists(zap_dir):
        os.makedirs(zap_dir, exist_ok=True)

    install_dir = os.path.dirname(jar_path)
    plugin_install_dir = os.path.join(install_dir, "plugin")
    plugin_zap_dir = os.path.join(zap_dir, "plugin")

    if not os.path.exists(plugin_zap_dir) and os.path.exists(plugin_install_dir):
        os.makedirs(plugin_zap_dir, exist_ok=True)
        for f in os.listdir(plugin_install_dir):
            src = os.path.join(plugin_install_dir, f)
            dest = os.path.join(plugin_zap_dir, f)
            if not os.path.exists(dest):
                shutil.copy2(src, dest)

    java_options: List[str] = []

    if use_toml and config.get("JAVA_OPTIONS", {}).get("flags"):
        java_options.extend(config["JAVA_OPTIONS"]["flags"])
    else:
        java_options.append("-Xmx2g")

    java_options.append(f"-Djava.io.tmpdir={tmp_dir}")

    config_flags: List[str] = []

    if use_toml and config.get("CONFIG", {}).get("flags"):
        config_flags.extend(config["CONFIG"]["flags"])
    else:
        if api_key:
            config_flags.append(f"api.key={api_key}")
        else:
            config_flags.append("api.disablekey=true")
        config_flags.append("api.addrs.addr.name=.*")
        config_flags.append("api.addrs.addr.regex=true")

    cmd = [
        "java",
    ]
    cmd.extend(java_options)
    cmd.extend(
        [
            "-jar",
            jar_path,
            "-daemon",
            "-dir",
            zap_dir,
            "-installdir",
            install_dir,
            "-host",
            host,
            "-port",
            str(port),
        ]
    )

    for flag in config_flags:
        cmd.extend(["-config", flag])

    console.print(f"[blue]Starting ZAP daemon...[/blue]")
    console.print(f"[gray]JAR: {jar_path}[/gray]")
    console.print(f"[gray]Port: {port}[/gray]")
    console.print(f"[gray]Working dir: {workspace}[/gray]")
    if use_toml:
        console.print(f"[gray]Config: TOML ({toml})[/gray]")

    try:
        if os.name == "nt":
            DETACHED_PROCESS = 0x00000008
            subprocess.Popen(
                cmd,
                cwd=workspace,
                creationflags=DETACHED_PROCESS,
                stdout=subprocess.DEVNULL,
                stderr=subprocess.DEVNULL,
            )
            console.print(f"[green]ZAP daemon started on {host}:{port}[/green]")
        else:
            with open(os.path.join(workspace, "zap.log"), "w") as f:
                subprocess.Popen(
                    cmd,
                    cwd=workspace,
                    stdout=f,
                    stderr=subprocess.STDOUT,
                    start_new_session=True,
                )
            console.print(f"[green]ZAP daemon started on {host}:{port}[/green]")
            console.print(f"[blue]Logs: {os.path.join(workspace, 'zap.log')}[/blue]")

    except Exception as e:
        console.print(f"[red]Failed to start ZAP daemon: {e}[/red]")
        raise typer.Exit(1)


def stop_daemon(
    workspace: Optional[str] = typer.Option(
        None, "--workspace", "-w", help="ZAP working directory"
    ),
    name: str = typer.Option("zap-daemon", "--name", "-N", help="Process name"),
):
    """Stop ZAP daemon."""
    from ..workspace import get_workspace

    if not workspace:
        workspace = get_workspace()

    console.print(f"[blue]Stopping ZAP daemon...[/blue]")

    pid_file = os.path.join(workspace, f"{name}.pid")

    try:
        if os.name == "nt":
            result = subprocess.run(["tasklist"], capture_output=True, text=True)
            for line in result.stdout.split("\n"):
                if "java.exe" in line:
                    subprocess.run(["taskkill", "/F", "/IM", "java.exe"], check=False)
                    console.print("[green]ZAP daemon stopped[/green]")
                    return
        else:
            with open(pid_file, "r") as f:
                pid = int(f.read().strip())
            os.kill(pid, signal.SIGTERM)
            os.remove(pid_file)
            console.print("[green]ZAP daemon stopped[/green]")
            return
    except FileNotFoundError:
        pass
    except ProcessLookupError:
        pass
    except Exception as e:
        console.print(f"[red]Error stopping daemon: {e}[/red]")

    console.print("[yellow]No running daemon found[/yellow]")
